src/System.Xml.XmlSerializer/src/System/Xml/Serialization/Compilation.cs

```diff
@@ -305,21 +303,27 @@ public override int GetHashCode()
 
     internal class TempAssemblyCache
     {
-        private Hashtable _cache = new Hashtable();
+        private Dictionary<TempAssemblyCacheKey, TempAssembly> _cache = new Dictionary<TempAssemblyCacheKey, TempAssembly>();
 
         internal TempAssembly this[string ns, object o]
         {
-            get { return (TempAssembly)_cache[new TempAssemblyCacheKey(ns, o)]; }
+            get
+            {
+                TempAssembly tempAssembly;
+                _cache.TryGetValue(new TempAssemblyCacheKey(ns, o), out tempAssembly);
+                return  tempAssembly;
+            }
         }
 
         internal void Add(string ns, object o, TempAssembly assembly)
         {
             TempAssemblyCacheKey key = new TempAssemblyCacheKey(ns, o);
             lock (this)
             {
-                if (_cache[key] == assembly) return;
-                Hashtable clone = new Hashtable();
-                foreach (object k in _cache.Keys)
+                TempAssembly tempAssembly;
+                if (_cache.TryGetValue(key, out tempAssembly) && tempAssembly == assembly) return;
+                var clone = new Dictionary<TempAssemblyCacheKey, TempAssembly>();
+                foreach (var k in _cache.Keys)
```

Comment:
It'd be more efficient to do this:

```csharp
TempAssembly tempAssembly;
if (_cache.TryGetValue(key, out tempAssembly) && tempAssembly == assembly) return;
_cache = new Dictionary<TempAssemblyCacheKey, TempAssembly>(_cache); // clone
```

 - Avoids allocating the collection object for `Keys`
 - Avoids having to lookup the value from the key
 - Take's advantage of the optimization in `Dictionary<TKey, TValue>`'s "copy" constructor.